chew.go

```diff
@@ -201,3 +424,102 @@ func processURL(url string, ctxs ...context.Context) ([]common.Chunk, error) {
 
 	return processor(resp.Body, url)
 }
+
+func getRobotsTxtInfo(urlStr string) (bool, time.Duration, error) {
+	parsedURL, err := url.Parse(urlStr)
+	if err != nil {
+		return false, 0, err
+	}
+
+	robotsURL := fmt.Sprintf("%s://%s/robots.txt", parsedURL.Scheme, parsedURL.Host)
+
+	robotsMu.RLock()
+	robotsData, exists := robotsCache[robotsURL]
+	robotsMu.RUnlock()
+
+	/*
+		This bit of code makes a lot of very unhealthy assumptions, in an ideal world this
+		would be more straightforward but unfortunately it's not.
+
+		Use this with caution and respect for site owners.
+	*/
+	if !exists {
+		resp, err := http.Get(robotsURL)
+		if err != nil {
+			return true, crawlDelay, nil
+		}
+		defer resp.Body.Close()
+
+		robotsData, err = robotstxt.FromResponse(resp)
+		if err != nil {
+			return true, crawlDelay, nil
+		}
+
+		robotsMu.Lock()
+		robotsCache[robotsURL] = robotsData
+		robotsMu.Unlock()
+	}
+
+	allowed := robotsData.TestAgent(parsedURL.Path, userAgent)
+
+	return allowed, crawlDelay, nil
+}
+
+// respectCrawlDelay ensures that subsequent requests to the same domain respect the specified crawl delay.
+func respectCrawlDelay(urlStr string, delay time.Duration) error {
+	parsedURL, err := url.Parse(urlStr)
+	if err != nil {
+		return err
+	}
+
+	domain := parsedURL.Hostname()
+
+	lastAccessMu.Lock()
+	defer lastAccessMu.Unlock()
+
+	lastAccess, exists := lastAccessTimes[domain]
+	if exists {
+		timeToWait := time.Until(lastAccess.Add(delay))
+		if timeToWait > 0 {
+			time.Sleep(timeToWait)
+		}
+	}
+
+	lastAccessTimes[domain] = time.Now()
+	return nil
+}
+
+func processWithRetry(url string, ctx context.Context) ([]common.Chunk, error) {
```

Comment:
```suggestion
func processWithRetry(ctx context.Context, url string) ([]common.Chunk, error) {
```

the context is by convention always the first argument, the same way the error is always the last one